csharp/extractor/Semmle.Extraction.CSharp.Standalone/Runtime.cs

```diff
@@ -3,38 +3,107 @@
 using System.Runtime.InteropServices;
 using System.IO;
 using System.Linq;
+using System.Text.RegularExpressions;
+using Semmle.BuildAnalyser;
 using Semmle.Util;
 
 namespace Semmle.Extraction.CSharp.Standalone
 {
     /// <summary>
     /// Locates .NET Runtimes.
     /// </summary>
-    internal static class Runtime
+    internal partial class Runtime
     {
+        private const string netCoreApp = "Microsoft.NETCore.App";
+        private const string aspNetCoreApp = "Microsoft.AspNetCore.App";
+
+        private readonly IDotNet dotNet;
         private static string ExecutingRuntime => RuntimeEnvironment.GetRuntimeDirectory();
 
-        /// <summary>
-        /// Locates .NET Core Runtimes.
-        /// </summary>
-        private static IEnumerable<string> CoreRuntimes
+        public Runtime(IDotNet dotNet) => this.dotNet = dotNet;
+
+        internal sealed class RuntimeVersion : IComparable<RuntimeVersion>
         {
-            get
+            private readonly string dir;
+            private Version Version { get; }
+            private Version? PreviewVersion { get; } = null;
+            private bool IsPreview => PreviewVersion is not null;
+            public string FullPath
             {
-                var dotnetPath = FileUtils.FindProgramOnPath(Win32.IsWindows() ? "dotnet.exe" : "dotnet");
-                var dotnetDirs = dotnetPath is not null
-                    ? new[] { dotnetPath }
-                    : new[] { "/usr/share/dotnet", @"C:\Program Files\dotnet" };
-                var coreDirs = dotnetDirs.Select(d => Path.Combine(d, "shared", "Microsoft.NETCore.App"));
+                get
+                {
+                    var preview = IsPreview ? $"-preview.{PreviewVersion}" : "";
+                    var version = Version + preview;
+                    return Path.Combine(dir, version);
+                }
+            }
 
-                var dir = coreDirs.FirstOrDefault(Directory.Exists);
-                if (dir is not null)
+            private static Version MakeVersion(string version)
+            {
+                var versionParts = version.Split('.');
+                return new Version(int.Parse(versionParts[0]), int.Parse(versionParts[1]), int.Parse(versionParts[2]));
```

Comment:
You can use `Version.Parse(version)` instead of manually parsing the components.